packages/codecs-data-structures/src/tuple.ts

```diff
@@ -100,17 +100,20 @@ export function getTupleDecoder<TTo extends AnyArray>(items: WrapInDecoder<[...T
  *
  * @param items - The codecs to use for each item in the tuple.
  */
-export function getTupleCodec<TFrom extends AnyArray, TTo extends TFrom = TFrom>(
-    items: WrapInFixedSizeCodec<[...TFrom], [...TTo]>,
-): FixedSizeCodec<TFrom, TTo>;
-export function getTupleCodec<TFrom extends AnyArray, TTo extends TFrom = TFrom>(
-    items: WrapInCodec<[...TFrom], [...TTo]>,
-): VariableSizeCodec<TFrom, TTo>;
-export function getTupleCodec<TFrom extends AnyArray, TTo extends TFrom = TFrom>(
-    items: WrapInCodec<[...TFrom], [...TTo]>,
-): Codec<TFrom, TTo> {
+export function getTupleCodec<const TItems extends readonly FixedSizeCodec<any>[]>(
+    items: TItems,
+): FixedSizeCodec<GetEncoderTypeFromItems<TItems>, GetDecoderTypeFromItems<TItems> & GetEncoderTypeFromItems<TItems>>;
+export function getTupleCodec<const TItems extends readonly Codec<any>[]>(
+    items: TItems,
+): VariableSizeCodec<
+    GetEncoderTypeFromItems<TItems>,
+    GetDecoderTypeFromItems<TItems> & GetEncoderTypeFromItems<TItems>
+>;
+export function getTupleCodec<const TItems extends readonly Codec<any>[]>(
+    items: TItems,
+): Codec<GetEncoderTypeFromItems<TItems>, GetDecoderTypeFromItems<TItems> & GetEncoderTypeFromItems<TItems>> {
     return combineCodec(
-        getTupleEncoder(items as WrapInEncoder<[...TFrom]>),
-        getTupleDecoder(items as WrapInDecoder<[...TTo]>),
+        getTupleEncoder(items),
+        getTupleDecoder(items) as Decoder<GetDecoderTypeFromItems<TItems> & GetEncoderTypeFromItems<TItems>>,
```

Comment:
I know you do this in other codecs, but why does the decoder type need both types in here? I would assume `GetDecoderTypeFromItems<TItems>>` would suffice.

Reply:
Because `Codec<TFrom, TTo>` has the following type parameter constraint: `TTo extends TFrom`. That's because the encoded type can be _looser_ than the decoded type. For instance `getU64Codec` returns a `Codec<number | bigint, bigint>` meaning you can either pass a number or a bigint to encode but when decoding, you'll always get a bigint.

Here we know that the Codec will satisfy this constraint because the nested codecs must also satisfy this constraint. But TypeScript doesn't know that so by telling it that it also satisfies the `TFrom` type, we essentially tell TypeScript that the created codec is valid.

Reply:
Gotcha, thanks! Was just curious.